Add version number to bytecode.

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36439 Add version number to bytecode.**

A proposal of versioning in bytecode, as suggested by @dzhulgakov in the internal post: https://fb.workplace.com/groups/pytorch.mobile.work/permalink/590192431851054/

kProducedBytecodeVersion is added. If the model version is not the same as the number in the code, an error will be thrown. 

The updated bytecode would look like below. It's a tuple of elements, where the first element is the version number.
```
(3,
 ('__torch__.m.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::Int', 'Tensor'),)),
   ('constants', ()),
   ('types', ()),
   ('register_size', 2))))
```

Differential Revision: [D22433532](https://our.internmc.facebook.com/intern/diff/D22433532)

iseeyuan committed:
Add version number to bytecode.

[ghstack-poisoned]

iseeyuan committed:
Update on "Add version number to bytecode."




[ghstack-poisoned]

dzhulgakov commented on torch/csrc/jit/serialization/import_export_constants.h:
```diff
@@ -1,10 +1,13 @@
 constexpr size_t BYTECODE_INDEX_CONSTANT = 2;
 constexpr size_t BYTECODE_INDEX_TYPE = 3;
+constexpr char BYTECODE_VERSION_STR[] = "version";
+constexpr int64_t BYTECODE_VERSION_NUMBER = 1;
```
do you expect to only support only one version at any point? Otherwise we can have Minimum/Producer/Maximum version similarly to how we do for main torchscript serialization

iseeyuan replied:
@zdevito mentioned about Minimum/Producer/Maximum version in main container. Sounds a good idea to me and I'll follow up this route.

zdevito commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -277,6 +278,7 @@ class ScriptModuleSerializer {
 
   void writeByteCode(const Module& module) {
     std::vector<c10::IValue> elements;
+    elements.emplace_back(Tup({BYTECODE_VERSION_STR, BYTECODE_VERSION_NUMBER}));
```
This format it weird to me because it takes version with a key but doesn't take the list of elements with a key (and then inlines the list of elements into a tuple with the version tuple. Is this suppose to be a named table of things? or a list of things? This mixes both.
`(("version", version), <elements>)`?

zdevito reviewed (commented):
I have a vague concern about adding more than one version number for every sub-part of the format. We had to go through effort to _remove_ duplicated version numbers. My proposal is to use the container format version number for the bytecode as well. To do this, each commit will have the following:

```
kMinSupportedFileFormatVersion = ...
// already exists, stays the same, technically because of BC this never gets changed
kMaxSupportedFileFormatVersion = ...
// already exists, stays the same

kProducedFileFormatVersion = ...
// the version we write for standard torchscript without bytecode

kProducedBytecodeVersion = ...
// the version we write when the archive contains bytecode. Possibly _higher_ than kProducedFileFormatVersion.
```

If we make a change to the bytecode that doesn't affect non-bytecode, we update kProducedBytecodeVersion _and_ kMaxSupportedFileFormatVersion to the new number. There are no forward compatibility problems introduced for non-bytecode files, because these still get written with the old number (which can be read by either the new or old releases). When someone updates kProducedFileFormatVersion, they will also update kProducedBytecodeVersion to the same version.

iseeyuan commented:
> I have a vague concern about adding more than one version number for every sub-part of the format. We had to go through effort to _remove_ duplicated version numbers. My proposal is to use the container format version number for the bytecode as well. To do this, each commit will have the following:
> 
> ```
> kMinSupportedFileFormatVersion = ...
> // already exists, stays the same, technically because of BC this never gets changed
> kMaxSupportedFileFormatVersion = ...
> // already exists, stays the same
> 
> kProducedFileFormatVersion = ...
> // the version we write for standard torchscript without bytecode
> 
> kProducedBytecodeVersion = ...
> // the version we write when the archive contains bytecode. Possibly _higher_ than kProducedFileFormatVersion.
> ```
> 
> If we make a change to the bytecode that doesn't affect non-bytecode, we update kProducedBytecodeVersion _and_ kMaxSupportedFileFormatVersion to the new number. There are no forward compatibility problems introduced for non-bytecode files, because these still get written with the old number (which can be read by either the new or old releases). When someone updates kProducedFileFormatVersion, they will also update kProducedBytecodeVersion to the same version.

Thanks for the suggestion! Sounds a great idea. One question on BC for bytecode. It sounds more difficult to maintain BC for bytecode than the standard torchscript. For standard torchscript do we compare A) kMinSupportedFileFormatVersion <= kProducedFileFormatVersion <= kProducedFileFormatVersion. For bytecode, should we do B) kMinSupportedFileFormatVersion <= kProducedBytecodeVersion <= kProducedFileFormatVersion, in addition to A)? I'm wondering if kMinSupportedFileFormatVersion is "too old" for bytecode if there are some bytecode only changes?

Commit:
Update on "Add version number to bytecode."


A proposal of versioning in bytecode, as suggested by @dzhulgakov in the internal post: https://fb.workplace.com/groups/pytorch.mobile.work/permalink/590192431851054/

The updated bytecode would look like:
```
(('version', 1),
 ('__torch__.m.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::Int', 'Tensor'),)),
   ('constants', ()),
   ('types', ()),
   ('register_size', 2))))
```

Commit:
Update on "Add version number to bytecode."


A proposal of versioning in bytecode, as suggested by @dzhulgakov in the internal post: https://fb.workplace.com/groups/pytorch.mobile.work/permalink/590192431851054/

kProducedBytecodeVersion is added. If the model version is not the same as the number in the code, an error will be thrown. 

The updated bytecode would look like below. It's a tuple of elements, where the first element is the version number.
```
(3,
 ('__torch__.m.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::Int', 'Tensor'),)),
   ('constants', ()),
   ('types', ()),
   ('register_size', 2))))
```

bhosmer commented on torch/csrc/jit/serialization/import_export_constants.h:
```diff
@@ -1,4 +1,5 @@
 #pragma once
+#include <string>
```
I don't think this is the right place for this :) but also not sure what the purpose is...

iseeyuan replied:
The major purpose for BYTECODE_INDEX_... constants, is that they are used both in exporting module and exporting op names. As you suggested in my previous PR, I replaced the inline numbers with those constants.

bhosmer replied:
Right, but I'd move this include out to the source files that use it, rather than including it here - it's not needed for the definitions in this file.

bhosmer commented on caffe2/serialize/inline_container.h:
```diff
@@ -99,6 +99,10 @@ constexpr uint64_t kMaxSupportedFileFormatVersion = 0x3L;
 
+// the version we write when the archive contains bytecode.
+// Possibly _higher_ than kProducedFileFormatVersion.
+constexpr uint64_t kProducedBytecodeVersion = 0x3L;
```
Also worth specifying (in a comment) the relationship between `kProducedBytecodeVersion`, `kProducedFileFormatVersion` and `kMaxSupportedFileFormatVersion`.

bhosmer commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -277,6 +278,8 @@ class ScriptModuleSerializer {
   void writeByteCode(const Module& module) {
     std::vector<c10::IValue> elements;
+    elements.emplace_back(
+        static_cast<int64_t>(caffe2::serialize::kProducedBytecodeVersion));
```
We're writing a version number for every module? We want a single version for the whole saved model, right? 

Couple followups:
1. Where does `kProducedFileFormatVersion` get saved?
2. When we save bytecode, do both `kProducedFileFormatVersion` and `kProducedBytecodeVersion` get saved, or just the latter?

iseeyuan replied:
Yes we have single version for the whole saved model. writeByteCode is called inside serialize(), which is called only once for the top-level module. 
1. It's saved in at the beginning of the bytecode section. Please refer to the summary of this PR for the example.
2. Only the latter is saved.

bhosmer replied:
Got it, thanks!

bhosmer reviewed (commented):
Hey, sorry for the delay! Had a few comments, most are simple and we can follow up on all of them this afternoon.

iseeyuan committed:
Update on "Add version number to bytecode."


A proposal of versioning in bytecode, as suggested by @dzhulgakov in the internal post: https://fb.workplace.com/groups/pytorch.mobile.work/permalink/590192431851054/

kProducedBytecodeVersion is added. If the model version is not the same as the number in the code, an error will be thrown. 

The updated bytecode would look like below. It's a tuple of elements, where the first element is the version number.
```
(3,
 ('__torch__.m.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::Int', 'Tensor'),)),
   ('constants', ()),
   ('types', ()),
   ('register_size', 2))))
```

[ghstack-poisoned]

iseeyuan committed:
Update on "Add version number to bytecode."


A proposal of versioning in bytecode, as suggested by @dzhulgakov in the internal post: https://fb.workplace.com/groups/pytorch.mobile.work/permalink/590192431851054/

kProducedBytecodeVersion is added. If the model version is not the same as the number in the code, an error will be thrown. 

The updated bytecode would look like below. It's a tuple of elements, where the first element is the version number.
```
(3,
 ('__torch__.m.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::Int', 'Tensor'),)),
   ('constants', ()),
   ('types', ()),
   ('register_size', 2))))
```

[ghstack-poisoned]

iseeyuan committed:
Update on "Add version number to bytecode."


A proposal of versioning in bytecode, as suggested by @dzhulgakov in the internal post: https://fb.workplace.com/groups/pytorch.mobile.work/permalink/590192431851054/

kProducedBytecodeVersion is added. If the model version is not the same as the number in the code, an error will be thrown. 

The updated bytecode would look like below. It's a tuple of elements, where the first element is the version number.
```
(3,
 ('__torch__.m.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::Int', 'Tensor'),)),
   ('constants', ()),
   ('types', ()),
   ('register_size', 2))))
```

[ghstack-poisoned]

iseeyuan committed:
Update on "Add version number to bytecode."


A proposal of versioning in bytecode, as suggested by @dzhulgakov in the internal post: https://fb.workplace.com/groups/pytorch.mobile.work/permalink/590192431851054/

kProducedBytecodeVersion is added. If the model version is not the same as the number in the code, an error will be thrown. 

The updated bytecode would look like below. It's a tuple of elements, where the first element is the version number.
```
(3,
 ('__torch__.m.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::Int', 'Tensor'),)),
   ('constants', ()),
   ('types', ()),
   ('register_size', 2))))
```

[ghstack-poisoned]

bhosmer commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -87,7 +84,28 @@ void print_unsupported_ops_and_throw(
     const std::vector<IValue>& vals,
     mobile::CompilationUnit& mcu) {
-  for (const auto& element : vals) {
+  TORCH_CHECK(
```
If we're supporting old models (per lines 90-98, below) then I don't think we want this check, right?

iseeyuan replied:
Updated the message to just check the size to avoid access error later.

bhosmer reviewed (approved):
Looks good, thanks especially for the good comments and messages! One minor suggestion inline.

iseeyuan committed:
Update on "Add version number to bytecode."


A proposal of versioning in bytecode, as suggested by @dzhulgakov in the internal post: https://fb.workplace.com/groups/pytorch.mobile.work/permalink/590192431851054/

kProducedBytecodeVersion is added. If the model version is not the same as the number in the code, an error will be thrown. 

The updated bytecode would look like below. It's a tuple of elements, where the first element is the version number.
```
(3,
 ('__torch__.m.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('MOVE', 2, 0),
     ('OP', 0, 0),
     ('RET', 0, 0))),
   ('operators', (('aten::Int', 'Tensor'),)),
   ('constants', ()),
   ('types', ()),
   ('register_size', 2))))
```

[ghstack-poisoned]
